src/messages/mod.rs

```diff
@@ -37,62 +35,114 @@ use xor_name::{Prefix, XorName};
     /// The body of the message.
-    pub variant: Variant,
+    variant: Variant,
     /// Source's knowledge of the destination section key. If present, the destination can use it
     /// to determine the length of the proof of messages sent to the source so the source would
     /// trust it (the proof needs to start at this key).
-    pub dst_key: Option<bls::PublicKey>,
-}
-
-/// Partially deserialized message.
-#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
-pub struct PartialMessage {
-    /// Destination location.
-    pub dst: DstLocation,
-}
+    dst_key: Option<bls::PublicKey>,
 
-impl PartialMessage {
-    /// Deserialize the message.
-    pub fn from_bytes(bytes: &Bytes) -> Result<Self> {
-        Ok(bincode::deserialize(&bytes[..])?)
-    }
+    /// Serialised message, this is a signed and fully serialised message ready to send.
+    #[serde(skip)]
+    serialized: Bytes,
+    #[serde(skip)]
+    hash: MessageHash,
 }
 
 impl Message {
-    /// Deserialize the message.
+    /// Deserialize the message. Only called on message receipt.
+    #[allow(unused)] // prefix field not used below
     pub(crate) fn from_bytes(bytes: &Bytes) -> Result<Self> {
-        Ok(bincode::deserialize(&bytes[..])?)
+        let mut msg: Message = bincode::deserialize(&bytes[..])?;
+        let signed_bytes = serialize_for_signing(&msg.dst, msg.dst_key.as_ref(), &msg.variant)?;
+        match msg.src.clone() {
+            SrcAuthority::Node {
+                public_id,
+                signature,
+            } => {
+                if public_id.verify(&signed_bytes, &signature) {
+                    msg.serialized = bytes.clone();
+                    msg.hash = MessageHash::from_bytes(bytes);
+                    Ok(msg)
+                } else {
+                    Err(RoutingError::FailedSignature)
+                }
+            }
+            SrcAuthority::Section {
+                prefix,
+                signature,
+                proof,
+            } => {
+                // FIXME Assumes the nodes proof last key is the one signing this message
+                if proof.last_key().verify(&signature, &signed_bytes) {
+                    msg.serialized = bytes.clone();
+                    msg.hash = MessageHash::from_bytes(bytes);
+                    Ok(msg)
+                } else {
+                    Err(RoutingError::FailedSignature)
+                }
+            }
+        }
     }
 
-    /// Serialize the message.
-    pub(crate) fn to_bytes(&self) -> Result<Bytes> {
-        Ok(bincode::serialize(self)?.into())
+    /// send across wire
+    pub(crate) fn to_bytes(&self) -> Bytes {
+        self.serialized.clone()
+    }
+
+    /// Creates a signed message where signature is assumed valid.
+    pub fn new_signed(
+        dst: DstLocation,
+        src: SrcAuthority,
+        variant: Variant,
+        dst_key: Option<bls::PublicKey>,
+    ) -> Result<Message> {
+        let mut msg = Message {
```

Comment:
better use a construct to populate serialized & hash field internally? which can avoid using default as well?

Reply:
Yes, I played a lot with this, sub classes, options, and more. The default() trick is a serde thing. So it deserialised to the type by populating the ignored fields with Default::default() 

I am not sure this is the best route yet, but it's solid and allows us to serialise ignoring these fields. It started as a mechanism to memoize them via options, but that messed up a lot of code that read the values a lot. So many match statements of map deconstructs. So I ended up with this. It may be we end up with a custom serialise/deserialize but it feels a bit messier doing that. 
I am sure this will improve over time but for now I want to create a barrier that if a message sxists then it's signed correctly with the key supplied in src. Still have to consider the section message being sperate and see if we can bring it an and then consider validation again to be sig + recodnised.

Reply:
Ah, ok, I get it. Thx for the clarification.